File: shared/type_strategy.ts
/**
 * TypeScript Type System Strategy for FormMonkey
 * 
 * This file documents the approach for managing TypeScript types in the FormMonkey frontend,
 * with a focus on aligning with the shared type system.
 */

/**
 * DeprecationWarning is used to mark types that should no longer be used
 * and should be replaced with imports from the shared type system.
 */
export type DeprecationWarning<T> = T & {
  /** @deprecated This type is duplicated and should be imported from shared/types instead */
  _deprecationWarning?: never;
};

/**
 * Mark a type as deprecated with a migration path
 * This helps developers understand how to migrate from local types to shared types
 */
export function markAsDeprecated<T>(typeName: string): string {
  return `
/**
 * @deprecated This type is duplicated from shared types.
 * Please import { ${typeName} } from '@shared/types' instead.
 */`;
}

/**
 * Type Compatibility Guide
 * 
 * When working with types that exist in both TypeScript and Python,
 * follow these guidelines to ensure cross-platform compatibility:
 * 
 * 1. Use camelCase for all property names in TypeScript
 * 2. Ensure corresponding Python models use snake_case
 * 3. Add Zod schemas with identical validation rules as Pydantic
 * 4. Include JSDoc comments that match Python docstrings
 * 5. Use type aliases (export type X = Y) for better refactoring support
 * 6. Avoid using "any" - prefer unknown with type guards
 * 7. Use discriminated unions for polymorphic types
 */

/**
 * Type Migration Registry
 * 
 * This tracks the status of type migrations from local implementations to shared types
 */
export enum TypeMigrationStatus {
  Pending = 'pending',
  InProgress = 'in_progress',
  Completed = 'completed',
  Verified = 'verified'
}

export const TYPE_MIGRATION_REGISTRY = {
  // Core types
  DocumentMetadata: TypeMigrationStatus.Completed,
  UserProfile: TypeMigrationStatus.Completed,
  FileMetadata: TypeMigrationStatus.Completed,
  UploadMetadata: TypeMigrationStatus.Completed,
  
  // Enums
  ProcessingStatus: TypeMigrationStatus.Completed,
  FileStatus: TypeMigrationStatus.Completed,
  UploadStatus: TypeMigrationStatus.Completed,
  FieldType: TypeMigrationStatus.Completed,
  PredictionSource: TypeMigrationStatus.Completed,
  CorrectionReason: TypeMigrationStatus.Completed,
  
  // AI-related types (Phase 4)
  AIPredictedField: TypeMigrationStatus.Completed,
  UserCorrection: TypeMigrationStatus.Completed,
  AlternativePrediction: TypeMigrationStatus.Completed,
  
  // TODO: Add remaining types with their migration status
};

File: shared/type_strategy.py
"""
Type System Strategy for FormMonkey

This file documents the architectural approach for managing types across the FormMonkey platform.
It provides guidance on how to maintain type consistency between frontend (TypeScript) 
and backend (Python) implementations.
"""

from enum import Enum

class TypeMigrationStatus(str, Enum):
    """Tracks the migration status of types in the system."""
    PENDING = "pending"               # Type needs to be migrated to shared types
    IN_PROGRESS = "in_progress"       # Migration in progress
    COMPLETED = "completed"           # Migration complete
    VERIFIED = "verified"             # Migration verified with tests

# Type migration registry
# This documents the status of type migrations from local implementations to shared types
TYPE_MIGRATION_REGISTRY = {
    # Core types
    "DocumentMetadata": TypeMigrationStatus.COMPLETED,
    "UserProfile": TypeMigrationStatus.COMPLETED,
    "FileMetadata": TypeMigrationStatus.COMPLETED,
    "UploadMetadata": TypeMigrationStatus.COMPLETED,
    
    # Enums
    "ProcessingStatus": TypeMigrationStatus.COMPLETED,
    "FileStatus": TypeMigrationStatus.COMPLETED,
    "UploadStatus": TypeMigrationStatus.COMPLETED,
    "FieldType": TypeMigrationStatus.COMPLETED,
    "PredictionSource": TypeMigrationStatus.COMPLETED,
    "CorrectionReason": TypeMigrationStatus.COMPLETED,
    
    # AI-related types (Phase 4)
    "AIPredictedField": TypeMigrationStatus.COMPLETED,
    "UserCorrection": TypeMigrationStatus.COMPLETED,
    "AlternativePrediction": TypeMigrationStatus.COMPLETED,
    
    # TODO: Add remaining types with their migration status
}

class DeprecationStrategy:
    """Strategy for deprecating duplicate types."""
    
    @staticmethod
    def mark_deprecated(module_name: str, type_name: str) -> str:
        """Generate a deprecation notice for a type."""
        return f"""
        # DEPRECATED: This type is deprecated and will be removed in a future version.
        # Import from shared.types instead: from shared.types import {type_name}
        """
    
    @staticmethod
    def generate_migration_guide(type_name: str) -> str:
        """Generate migration guide for a type."""
        return f"""
        Migration Guide for {type_name}:
        
        1. Replace imports:
           - Frontend: import {{ {type_name} }} from '@shared/types';
           - Backend: from shared.types import {type_name}
           
        2. Update field names:
           - Frontend uses camelCase
           - Backend uses snake_case
           
        3. Ensure validation rules match between:
           - Zod schema in types.ts
           - Pydantic model in types.py
        """
